### PR TITLE
Implement APIs for some threading metrics (CoreCLR)

### PR DESCRIPTION
**Pull request by kouvel**
- API review: https://github.com/dotnet/corefx/issues/35500

### CODE REVIEW

**Comment by kouvel** [comment]
There was no measurable change to perf from the changes

File: src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs [hunk 1]
```diff
@@ -383,6 +383,26 @@ public object TrySteal(ref bool missedSteal)
...
             }
+
+            public int Count
+            {
+                get
+                {
+                    bool lockTaken = false;
+                    try
+                    {
+                        m_foreignLock.Enter(ref lockTaken);
```
Comment (stephentoub) [inline src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs]: Do we need to lock, or could we just say that the count is an estimate and hot read the head and tail?

File: src/vm/syncblk.cpp [hunk 2]
```diff
@@ -3049,6 +3051,7 @@ BOOL AwareLock::EnterEpilogHelper(Thread* pCurThread, INT32 timeOut)
     FireEtwContentionStart_V1(ETW::ContentionLog::ContentionStructs::ManagedContention, GetClrInstanceId());
 
     LogContention();
+    InterlockedIncrement64((LONGLONG *)&s_lockContentionCount);
```
Comment (stephentoub) [inline src/vm/syncblk.cpp]: Isn't this going to make contention even more expensive, creating contention not only on this individual monitor but synchronizing it with every other monitor experiencing contention as well, at least a little?

File: src/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.cs [hunk 3]
```diff
@@ -228,6 +228,36 @@ public static void GetAvailableThreads(out int workerThreads, out int completion
...
+        /// <summary>
+        /// Gets the number of work items that have been processed so far.
+        /// </summary>
+        /// <remarks>
+        /// For a thread pool implementation that may have different types of work items, the count includes all types.
+        /// </remarks>
+        public static extern long CompletedWorkItemCount
+        {
+            [MethodImpl(MethodImplOptions.InternalCall)]
+            get;
```
Comment (stephentoub) [inline src/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.cs]: Interesting; I wasn't expecting to see this implemented in native.  Why is that necessary?  I'm also missing where this is factoring in work items dequeued and executed from the managed queues.  Where does that cause the count to be incremented?

File: src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs (hunk 1 above)
Comment (kouvel) [reply src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs]: I was thinking about that but in the overflow case of the tail index the hot reads may yield a totally wrong result if read during the update. I figured perf is not that big an issue so may as well do the safe thing.

File: src/vm/syncblk.cpp (hunk 2 above)
Comment (kouvel) [reply src/vm/syncblk.cpp]: This thread will have to wait for the lock to be released, but the increment itself may not contend with other threads. It can add a little bit of overhead but we're about to spend several microseconds for context switching anyway so it should be negligible. It is still extra work though and this can add a little bit of CPU time in cases of frequent contention.

File: src/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.cs (hunk 3 above)
Comment (kouvel) [reply src/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.cs]: Timer, wait, and IO completion callbacks are initiated from the native side and there is already thread-local-ish code there to track completion counts for hill climbing, so I just built on top of that.

Comment (kouvel) [reply src/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.cs]: For work items in the managed queues this call increments the count:
https://github.com/dotnet/coreclr/blob/a006c44d06d56829a230bbbce17b74b21a0b8842/src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs#L647-L652

Here:
https://github.com/dotnet/coreclr/blob/a006c44d06d56829a230bbbce17b74b21a0b8842/src/vm/win32threadpool.h#L842-L847

File: src/vm/syncblk.cpp (hunk 2 above)
Comment (jkotas) [reply src/vm/syncblk.cpp]: All cores will start fighting over this cache line if there is a bit contention (that can be even spread over multiple locks). Each of these fights can cost 1000s cycles on many core systems, and the amplify the effect of contention that would be harmless otherwise.

I would recommend testing this on server class hardware with 32+ cores.

Comment (kouvel) [reply src/vm/syncblk.cpp]: CoreCLR is using thread-locals for tracking thread pool work item completion counts, maybe that is the reason as well and all of those should be switched. Will try it out on a larger machine.

**Comment by kouvel** [comment]
Some perf numbers:
- Left score = Before any changes, work items per ms
- Right score = After any changes, work items per ms

Initial change:

```
CoreCLR 4-core 4-thread 1-node          Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi   670.88 ±0.20%   677.22 ±0.16%      0.95%
TaskBurstWorkThroughput 1PcT 001.0PcWi  1642.87 ±0.39%  1661.91 ±0.45%      1.16%
TaskBurstWorkThroughput 1PcT 004.0PcWi  1859.70 ±0.61%  1877.41 ±0.63%      0.95%
TaskBurstWorkThroughput 1PcT 016.0PcWi  1696.58 ±0.60%  1681.65 ±0.57%     -0.88%
TaskBurstWorkThroughput 1PcT 064.0PcWi  1843.14 ±0.70%  1872.47 ±0.50%      1.59%
TaskBurstWorkThroughput 1PcT 256.0PcWi  2025.88 ±0.62%  2015.75 ±0.54%     -0.50%
TaskSustainedWorkThroughput 1PcT        2906.67 ±0.25%  2913.72 ±0.76%      0.24%
--------------------------------------  --------------  --------------  ---------

CoreRT 4-core 4-thread 1-node           Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi   403.73 ±0.21%   407.09 ±0.12%      0.83%
TaskBurstWorkThroughput 1PcT 001.0PcWi   976.50 ±0.11%   981.36 ±0.08%      0.50%
TaskBurstWorkThroughput 1PcT 004.0PcWi  2913.15 ±0.07%  2892.59 ±0.10%     -0.71%
TaskBurstWorkThroughput 1PcT 016.0PcWi  2610.12 ±0.06%  2589.73 ±0.08%     -0.78%
TaskBurstWorkThroughput 1PcT 064.0PcWi  3057.35 ±0.11%  3038.80 ±0.10%     -0.61%
TaskBurstWorkThroughput 1PcT 256.0PcWi  3235.46 ±0.11%  3207.21 ±0.13%     -0.87%
TaskSustainedWorkThroughput 1PcT        3943.24 ±0.17%  3924.89 ±0.08%     -0.47%
--------------------------------------  --------------  --------------  ---------

CoreCLR 48-core 48-thread 4-node        Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi   878.83 ±0.40%   877.38 ±0.40%     -0.16%
TaskBurstWorkThroughput 1PcT 001.0PcWi  4541.56 ±0.74%  4647.15 ±0.90%      2.32%
TaskBurstWorkThroughput 1PcT 004.0PcWi  5295.10 ±0.30%  5431.67 ±0.33%      2.58%
TaskBurstWorkThroughput 1PcT 016.0PcWi  5301.83 ±0.26%  5356.74 ±0.44%      1.04%
TaskBurstWorkThroughput 1PcT 064.0PcWi  5284.73 ±0.22%  5321.00 ±0.39%      0.69%
TaskBurstWorkThroughput 1PcT 256.0PcWi  5323.88 ±0.10%  5264.35 ±0.39%     -1.12%
TaskSustainedWorkThroughput 1PcT        7874.07 ±0.88%  8029.77 ±0.66%      1.98%
--------------------------------------  --------------  --------------  ---------

CoreRT 48-core 48-thread 4-node         Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi  3546.97 ±0.12%  3264.71 ±0.17%     -7.96%
TaskBurstWorkThroughput 1PcT 001.0PcWi  2133.60 ±0.23%  1823.23 ±0.49%    -14.55%
TaskBurstWorkThroughput 1PcT 004.0PcWi  4514.30 ±0.30%  3846.87 ±0.12%    -14.78%
TaskBurstWorkThroughput 1PcT 016.0PcWi  4867.67 ±0.29%  4115.50 ±0.21%    -15.45%
TaskBurstWorkThroughput 1PcT 064.0PcWi  4927.05 ±0.32%  4144.38 ±0.35%    -15.89%
TaskBurstWorkThroughput 1PcT 256.0PcWi  4541.31 ±0.50%  3946.43 ±0.45%    -13.10%
TaskSustainedWorkThroughput 1PcT        7061.97 ±0.41%  6064.36 ±0.37%    -14.13%
--------------------------------------  --------------  --------------  ---------
```

After change to use thread-locals for more things (monitor lock contention counting in CoreCLR, all of the relevant counting in CoreRT):

```
CoreCLR 4-core 4-thread 1-node          Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi   673.14 ±0.28%   672.85 ±0.25%     -0.04%
TaskBurstWorkThroughput 1PcT 001.0PcWi  1658.72 ±0.32%  1669.33 ±0.23%      0.64%
TaskBurstWorkThroughput 1PcT 004.0PcWi  1845.45 ±0.49%  1874.94 ±0.51%      1.60%
TaskBurstWorkThroughput 1PcT 016.0PcWi  1674.97 ±0.68%  1687.24 ±0.72%      0.73%
TaskBurstWorkThroughput 1PcT 064.0PcWi  1846.55 ±0.46%  1839.79 ±0.58%     -0.37%
TaskBurstWorkThroughput 1PcT 256.0PcWi  2024.60 ±0.35%  2034.10 ±0.31%      0.47%
TaskSustainedWorkThroughput 1PcT        2904.78 ±0.33%  3051.85 ±0.42%      5.06%
--------------------------------------  --------------  --------------  ---------

CoreRT 4-core 4-thread 1-node           Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi   406.80 ±0.25%   405.98 ±0.31%     -0.20%
TaskBurstWorkThroughput 1PcT 001.0PcWi   977.56 ±0.06%   977.15 ±0.09%     -0.04%
TaskBurstWorkThroughput 1PcT 004.0PcWi  2838.71 ±0.53%  2851.90 ±0.61%      0.46%
TaskBurstWorkThroughput 1PcT 016.0PcWi  2600.13 ±0.18%  2602.24 ±0.14%      0.08%
TaskBurstWorkThroughput 1PcT 064.0PcWi  3037.53 ±0.14%  3030.35 ±0.08%     -0.24%
TaskBurstWorkThroughput 1PcT 256.0PcWi  3204.87 ±0.32%  3222.19 ±0.13%      0.54%
TaskSustainedWorkThroughput 1PcT        3930.82 ±0.20%  3917.91 ±0.08%     -0.33%
--------------------------------------  --------------  --------------  ---------

CoreCLR 48-core 48-thread 4-node        Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi   874.75 ±0.33%   706.56 ±0.46%    -19.23%
TaskBurstWorkThroughput 1PcT 001.0PcWi  4472.33 ±1.28%  2294.10 ±1.53%    -48.70%
TaskBurstWorkThroughput 1PcT 004.0PcWi  5324.16 ±0.33%  5411.74 ±0.44%      1.64%
TaskBurstWorkThroughput 1PcT 016.0PcWi  5321.38 ±0.27%  5379.79 ±0.13%      1.10%
TaskBurstWorkThroughput 1PcT 064.0PcWi  5267.20 ±0.39%  5419.75 ±0.20%      2.90%
TaskBurstWorkThroughput 1PcT 256.0PcWi  5241.08 ±0.46%  5389.11 ±0.26%      2.82%
TaskSustainedWorkThroughput 1PcT        7535.42 ±0.43%  7880.48 ±0.70%      4.58%
--------------------------------------  --------------  --------------  ---------

CoreRT 48-core 48-thread 4-node         Left score      Right score     ∆ Score %
--------------------------------------  --------------  --------------  ---------
TaskBurstWorkThroughput 1PcT 000.5PcWi  3562.74 ±0.10%  3280.68 ±0.22%     -7.92%
TaskBurstWorkThroughput 1PcT 001.0PcWi  2111.26 ±0.44%  1758.21 ±0.38%    -16.72%
TaskBurstWorkThroughput 1PcT 004.0PcWi  4507.55 ±0.18%  3648.46 ±0.15%    -19.06%
TaskBurstWorkThroughput 1PcT 016.0PcWi  4856.60 ±0.29%  4007.77 ±0.14%    -17.48%
TaskBurstWorkThroughput 1PcT 064.0PcWi  4958.45 ±0.38%  4023.59 ±0.36%    -18.85%
TaskBurstWorkThroughput 1PcT 256.0PcWi  4543.73 ±0.50%  3875.68 ±0.46%    -14.70%
TaskSustainedWorkThroughput 1PcT        7106.51 ±0.36%  5828.59 ±0.48%    -17.98%
--------------------------------------  --------------  --------------  ---------
```

Summary:
- Updated task burst tests to scale a bit better, code is [here](https://github.com/dotnet/coreclr/pull/13670#issuecomment-325858065) and I had modified it to use a medium-length delay (RandomMediumDelay) for the numbers above
- None of these regressions or improvements are caused by either change (counting with interlocked or thread-local)
- According to profiles, perf changes appear to be in code that was not changed. Perhaps slight changes in timing are affecting where and how badly contention occurs.
- For monitor lock contention counting, I tried a test that has multiple threads wait on an already acquired lock, then when all threads reach a waiting state the lock is released, and this repeats. I did not see any noticeable contention from using interlocked counting for Monitor lock contentions before waiting.
- Decided to go with the thread-local approach for all of the relevant counting, as in the profiles it doesn't appear to be any worse than counting with interlocked operations, and it shouldn't be a cause of contention in the future. At the moment the difference doesn't seem to be noticeable. For thread pool, probably there are other bottlenecks that would all need to be improved to see a difference.

File: src/vm/syncblk.cpp (hunk 2 above)
Comment (kouvel) [reply src/vm/syncblk.cpp]: See latest comment about perf, I didn't notice a perf difference but I have switched to use thread-locals

**Comment by kouvel** [comment]
The RT implementation is broken, ignore those perf numbers, I'll fix and retest

**Comment by kouvel** [comment]
Perf numbers after fixes were similar to before, updated inline above

**Comment by kouvel** [comment]
@dotnet-bot test this please

**Comment by kouvel** [comment]
Ping for review please

**Comment by kouvel** [comment]
@dotnet-bot test Ubuntu arm Cross Checked crossgen_comparison Build and Test

**Comment by jkotas** [comment]
I have been waiting for API review to be done before looking at this.

**Comment by kouvel** [comment]
Oh ok, I'll ping that one as well

**Comment by kouvel** [comment]
The API review was approved. Couple of pending items is:
- Whether to expose `PendingLocalWorkItemCount` and `PendingGlobalWorkItemCount`. I have asked for more info on whether they would be useful, by default I'll go ahead and remove them
- Whether we want to expose `CompletedWorkItemCount` at all.
  - I think that's the most expensive one to track of the lot, as it involves a bit of extra work per work item
  - It is possible to mostly eliminate the extra cost for managed work items in CoreCLR and CoreRT. Work items that may have some extra cost are timer callbacks in CoreRT Windows and IO completion work items in CoreCLR Windows.
  - The info would enable easily determining work item throughput and whether the thread pool is stalled

**Comment by kouvel** [comment]
The perf difference from adding `CompletedWorkItemCount` seems to be very small, so for now I'll consider that it can reasonably be added

**Comment by kouvel** [comment]
I have updated to remove local/global versions of PendingWorkItemCount. They can be added in the future if they would be useful. I think all three PRs (CoreCLR, CoreRT, and CoreFX) are now ready for review. @jkotas / @stephentoub, would you be able to take a look?

File: src/vm/threads.cpp [hunk 4]
```diff
@@ -8341,22 +8367,56 @@ LONG Thread::GetTotalThreadPoolCompletionCount()
...
+        NOTHROW;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    UINT64 total;
+    if (g_fEEStarted) //make sure we actually have a thread store
+    {
+        // make sure up-to-date thread-local counts are visible to us
+        ::FlushProcessWriteBuffers();
```
Comment (jkotas) [inline src/vm/threads.cpp]: Why do we need to do this? We are getting a stale counts by design.

Comment (kouvel) [reply src/vm/threads.cpp]: It's probably not necessary. It was there before on the path where hill climbing gets the work item completion counts. For hill climbing too it is stale counts by design.

File: src/vm/threads.cpp [hunk 5]
```diff
@@ -8332,31 +8358,59 @@ LONG Thread::GetTotalThreadPoolCompletionCount()
...
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    UINT64 total;
+    if (g_fEEStarted) //make sure we actually have a thread store
```
Comment (jkotas) [inline src/vm/threads.cpp]: This condition looks like a left-over from desktop. I think it is fine to delete it.

**Comment by jkotas** [comment]
> Preserve flush for hill climbing

Why do you think we need the flush for hill climbing?

**Comment by kouvel** [comment]
> Why do you think we need the flush for hill climbing?

Hill climbing wants to know what the current counts are so that it can respond and having the flush may help it respond sooner. Some processors may take longer than others to sync those updates. I doubt that it would take long enough to sync those updates to be comparable to the hill climbing interval, I'm not sure about that. I don't think it will make a noticeable difference on recent processors, not sure about older ones.

**Comment by jkotas** [comment]
> having the flush may help it respond sooner.

At the cost of slowing down every processor in the system...

**Comment by kouvel** [comment]
The hill climbing interval is between 10 and 200 ms, higher when the thread count is already at the min number of threads. I'm not sure if the perf hit would be significant on those time scales. I also doubt that the flush would improve anything, I just kept it because I don't have all of the information on why it was there to begin with.

**Comment by kouvel** [comment]
Changed my mind again, I'll just remove it. If not having a flush would be an issue it likely would have shown up in a bunch of other places before.

**Comment by kouvel** [comment]
@dotnet-bot test Windows_NT x64 Checked CoreFX Tests

File: src/vm/threads.cpp [hunk 6]
```diff
@@ -8332,32 +8358,44 @@ LONG Thread::GetTotalThreadPoolCompletionCount()
...
-    if (g_fEEStarted) //make sure we actually have a thread store
-    {
-        // make sure up-to-date thread-local counts are visible to us
-        ::FlushProcessWriteBuffers();
+    // enumerate all threads, summing their local counts.
+    ThreadStoreLockHolder tsl;
 
-        // enumerate all threads, summing their local counts.
-        ThreadStoreLockHolder tsl;
+    UINT64 total = GetOverflowCount(overflowCount);
```
Comment (stephentoub) [inline src/vm/threads.cpp]: I guess getting the overflow count first and then iterating through the threads ensures we won't double-count a thread's stats if the thread is currently being removed?

Comment (kouvel) [reply src/vm/threads.cpp]: Getting the total count and updating the overflow count when a thread is removed, are both done inside the thread store lock, so the order shouldn't matter.

**Comment by kouvel** [comment]
Merged in https://github.com/dotnet/coreclr/pull/24113
